src/Symfony/Component/Filesystem/Watcher/INotifyWatcher.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Filesystem\Watcher;
+
+use Symfony\Component\Filesystem\Exception\IOException;
+
+/**
+ * @author Pierre du Plessis <pdples@gmail.com>
+ *
+ * @internal
+ */
+final class INotifyWatcher implements WatcherInterface
+{
+    public function watch($path, callable $callback, float $timeout = null): void
+    {
+        $inotifyInit = inotify_init();
+
+        if (false === $inotifyInit) {
+            throw new IOException('Unable initialize inotify.', 0, null, $path);
+        }
+
+        stream_set_blocking($inotifyInit, false);
+
+        $isDir = is_dir($path);
+        $watchers = [];
+
+        if ($isDir) {
+            $watchers[] = inotify_add_watch($inotifyInit, $path, IN_CREATE | IN_DELETE | IN_MODIFY);
+
+            foreach ($this->scanPath("$path/*") as $path) {
+                $watchers[] = inotify_add_watch($inotifyInit, $path, IN_CREATE | IN_DELETE | IN_MODIFY);
+            }
+        } else {
+            $watchers[] = inotify_add_watch($inotifyInit, $path, IN_MODIFY);
+        }
+
+        try {
+            $read = [$inotifyInit];
+            $write = null;
+            $except = null;
+            $tvSec = null === $timeout ? null : 0;
+            $tvUsec = null === $timeout ? null : $timeout * 1000;
```

Comment:
Having the user set milliseconds instead of seconds gives them greater flexibility.
The parameter is properly documented in the `Filesystem:watch` method for users to know they should be using milliseconds

Reply:
Well normalization to correct unit can (and I argue should) still be done in FileSystem::watch, would make job for anybody reading code of these watchers easier.

Reply:
It can't really be normalized in the `Filesystem::watch` method, as the implementations use different values.

In the `FileChangeWatcher`, `microtime` is used, which returns a value in seconds. So the timeout needs to be divided by 1000 to convert it into seconds in order to do the timeout calculation.

Here, the argument expects a value in microseconds, so we need to multiple the timeout by 1000 to get a value in microseconds

Reply:
`Filesystem::watch` decides which watcher to use, so it knows which watcher requires which units, normalization can absolutely be done there.

Reply:
But that is an implementation detail for the specific watcher. Why should the Filesystem class care how to format the timeout for each watcher?

This also makes the interface difficult to implement, since you might get a value in seconds in some cases, and value in microseconds in other cases. Each new watcher that is added, now can't decide how it wants to use the timeout, but an external class now needs to decide what value it wants to send to the watcher

Reply:
> Why should the Filesystem class care how to format the timeout for each watcher?

Because it instantiates them

> makes the interface difficult to implement

Timeout doesn't need to be defined by interface, there is no reason for it. And it also might not make sense to implement for each implementation. Filesystem:watch can pass via constructor concrete units as required by each implementation.

> Each new watcher that is added, now can't decide how it wants to use the timeout

What makes you say that? It's pretty clear for each implementation what unit it needs. I see it quite opposite. There are pretty much only 3 units used: seconds, milliseconds, microseconds. New implementations will use either of those and centralizing this logic in FileSystem will prevent unnecessary duplication of same normalization code in all future implementations.